### PR TITLE
[bazel] add build for functorch

### LINKED ISSUE #101469: [bazel] add functorch to bazel build
### 🚀 The feature, motivation and pitch

We should work towards allowing `torch.compile()` to be run from the bazel build.
The first step would be enabling `functorch` in the bazel build.

### Alternatives

_No response_

### Additional context

_No response_

cc @malfet @seemethere

### PR DESCRIPTION
**Pull request by vors**
Fixes #101469

### CODE REVIEW

File: test/test_bazel.py [hunk 1]
```diff
@@ -9,7 +9,20 @@
...
     assert torch.eq(torch.tensor([[1, 2, 3]]) + torch.tensor([[4, 5, 6]]), torch.tensor([[5, 7, 9]])).all()
 
+def test_simple_compile_ts() -> None:
+
+    def foo(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        a = torch.sin(x)
+        b = torch.cos(y)
+        return a + b
+
+    opt_foo1 = torch.compile(foo(), backend="ts")
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_bazel.py:22]: Just drop the backend entirely, there is no reason to test TS here.

File: BUILD.bazel [hunk 2]
```diff
@@ -1696,6 +1696,36 @@ pybind_extension(
     ],
 )
 
+cc_library(
+    name = "functorch",
+    hdrs = glob([
+        "functorch/csrc/dim/*.h",
+    ]),
+    srcs = glob([
+        "functorch/csrc/dim/*.cpp",
```
Comment (ezyang) [inline BUILD.bazel:1705]: FWIW, this doesn't actually have any files that are required for torch.compile

**Review by ezyang — APPROVED** [review]
sure why not

File: test/test_bazel.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_bazel.py:22]: Then just do eager or aot_eager

**Comment by eqy** [comment]
I believe this PR introduced a test that somehow conflicts with #95722 , producing something like
```
INFO: From Testing //:test_bazel:
==================== Test output for //:test_bazel:
WARNING: Logging before InitGoogleLogging() is written to STDERR
Warning: Loading nvfuser library failed with: Error in dlopen: libnvfuser_codegen.so: cannot open shared object file: No such file or directory (function LoadingNvfuserLibrary)
terminate called after throwing an instance of 'c10::Error'
  what():  Type c10::intrusive_ptr<ConvPackedParamsBase<2>, c10::detail::intrusive_target_default_null_type<ConvPackedParamsBase<2> > > could not be converted to any of the known types.
Exception raised from operator() at aten/src/ATen/core/jit_type.h:1809 (most recent call first):
(no backtrace available)
```

However, I don't see any usage of `ConvPackedParamsBase` (which looks to be mostly in the quantized cuDNN backend). in the actual test @ `test/_test_bazel.py`. Running `test/_test_bazel.py` locally doesn't produce any errors on my end either, but I'm not able to build with bazel. Is it possible that always compiling the quantized cuDNN backend (which is now done in #95722) requires some actual additional source files to be added?

CC @vors @ezyang @malfet as I really don't have any knowledge of how the bazel build works.

**Comment by eqy** [comment]
Here's an example (it fails during tests rather than during build): https://github.com/pytorch/pytorch/actions/runs/5352136782/jobs/9706759181?pr=95722

**Comment by eqy** [comment]
It would be in quantization, yes---I suspect it would be here: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/quantized/cudnn/ConvPrepack.cpp

**Comment by eqy** [comment]
Unfortunately I don't have CI access to the workers, we might need to ask @malfet or @ezyang for that

**Comment by malfet** [comment]
@eqy can we keep a discussion in a open PR, as this one was landed a few month back and does not introduce any new regressions. Also, one doesn't need an access to CI workers, as bazel builds are quite hermetic. Are you saying it's not reproducing on your end?

**Pull request closed without merging**